From 2bd52ac4f8d041a5ca02da1a23d32f03bfeafae9 Mon Sep 17 00:00:00 2001
From: liuzhensong <liuzhensong@loongson.cn>
Date: Fri, 26 Nov 2021 10:35:38 +0800
Subject: [PATCH 11/11] binutils: change flags args and bits

  bfd/elfnn-loongarch.c
  gas/config/tc-loongarch.c
  gdb/loongarch-tdep.c
  include/elf/loongarch.h
---
 bfd/elfnn-loongarch.c   |  2 +-
 gdb/loongarch-tdep.c    |  4 ++--
 include/elf/loongarch.h | 38 ++++++++++++++++++++++++--------------
 3 files changed, 27 insertions(+), 17 deletions(-)

diff --git a/bfd/elfnn-loongarch.c b/bfd/elfnn-loongarch.c
index 406bbcf32e0..44a58e7be5a 100644
--- a/bfd/elfnn-loongarch.c
+++ b/bfd/elfnn-loongarch.c
@@ -401,7 +401,7 @@ elfNN_loongarch_merge_private_bfd_data (bfd *ibfd, struct bfd_link_info *info)
     }
 
   /* Disallow linking different ABIs.  */
-  if ((out_flags ^ in_flags) & EF_LOONGARCH_ABI)
+  if (EF_LOONGARCH_ABI(out_flags ^ in_flags) & EF_LOONGARCH_ABI_MASK)
     {
       _bfd_error_handler (_("%pB: can't link different ABI object."), ibfd);
       goto fail;
diff --git a/gdb/loongarch-tdep.c b/gdb/loongarch-tdep.c
index 66578da2a40..804ec43e5b1 100644
--- a/gdb/loongarch-tdep.c
+++ b/gdb/loongarch-tdep.c
@@ -1516,8 +1516,8 @@ loongarch_gdbarch_init (struct gdbarch_info info, struct gdbarch_list *arches)
   if (info.abfd != NULL
       && bfd_get_flavour (info.abfd) == bfd_target_elf_flavour)
     {
-      tdep->ef_abi = elf_elfheader (info.abfd)->e_flags & EF_LOONGARCH_ABI;
-      gdb_assert (0 != (tdep->ef_abi & EF_LOONGARCH_ABI_MASK));
+      tdep->ef_abi = EF_LOONGARCH_ABI(elf_elfheader (info.abfd)->e_flags);
+      gdb_assert (0 != tdep->ef_abi);
 
     }
 
diff --git a/include/elf/loongarch.h b/include/elf/loongarch.h
index b4d801ae9bd..c85dd0e92b6 100644
--- a/include/elf/loongarch.h
+++ b/include/elf/loongarch.h
@@ -93,27 +93,37 @@ RELOC_NUMBER (R_LARCH_GNU_VTENTRY, 58)
 END_RELOC_NUMBERS (R_LARCH_count)
 
 /* Processor specific flags for the ELF header e_flags field.  */
-#define EF_LOONGARCH_ABI_ILP32	    0b01
-#define EF_LOONGARCH_ABI_LP64	    0b11
-#define EF_LOONGARCH_ABI_MASK	    0b11
+#define EF_LOONGARCH_ABI_LP64_SOFT_FLOAT	0b001
+#define EF_LOONGARCH_ABI_LP64_SINGLE_FLOAT	0b010
+#define EF_LOONGARCH_ABI_LP64_DOUBLE_FLOAT	0b011
 
-#define EF_LOONGARCH_FLOAT_ABI_SOFT    0b001100
-#define EF_LOONGARCH_FLOAT_ABI_SINGLE  0b001000
-#define EF_LOONGARCH_FLOAT_ABI_DOUBLE  0b000000
-#define EF_LOONGARCH_FLOAT_ABI_MASK    0b111100
+#define EF_LOONGARCH_ABI_ILP32_SOFT_FLOAT	0b101
+#define EF_LOONGARCH_ABI_ILP32_SINGLE_FLOAT	0b110
+#define EF_LOONGARCH_ABI_ILP32_DOUBLE_FLOAT	0b111
+
+#define EF_LOONGARCH_ABI_MASK	    		0x07U
+
+#define EF_LOONGARCH_ABI_ILP32_MASK	    	0b100
+
+#define EF_LOONGARCH_ABI_FLOAT_MASK	    	0x03U
+
+#define EF_LOONGARCH_ABI_SOFT_FLOAT_MASK	0b001
+#define EF_LOONGARCH_ABI_SINGLE_FLOAT_MASK	0b010
+#define EF_LOONGARCH_ABI_DOUBLE_FLOAT_MASK	0b011
+
+#define EF_LOONGARCH_ABI(abi)	(EF_LOONGARCH_ABI_MASK & (abi))
 
-#define EF_LOONGARCH_IS_LP64(abi) \
-  ((abi & EF_LOONGARCH_ABI_MASK) == EF_LOONGARCH_ABI_LP64)
 #define EF_LOONGARCH_IS_ILP32(abi) \
-  ((abi & EF_LOONGARCH_ABI_MASK) == EF_LOONGARCH_ABI_ILP32)
+  (EF_LOONGARCH_ABI(abi) && (EF_LOONGARCH_ABI(abi) & EF_LOONGARCH_ABI_ILP32_MASK))
+#define EF_LOONGARCH_IS_LP64(abi) \
+  (EF_LOONGARCH_ABI(abi) && (!(EF_LOONGARCH_ABI(abi) & EF_LOONGARCH_ABI_ILP32_MASK)))
 
 #define EF_LOONGARCH_IS_SOFT_FLOAT(abi) \
-  ((abi & EF_LOONGARCH_FLOAT_ABI_MASK) == EF_LOONGARCH_FLOAT_ABI_SOFT)
+  (!((EF_LOONGARCH_ABI(abi) & EF_LOONGARCH_ABI_FLOAT_MASK) ^ EF_LOONGARCH_ABI_SOFT_FLOAT_MASK))
 #define EF_LOONGARCH_IS_SINGLE_FLOAT(abi) \
-  ((abi & EF_LOONGARCH_FLOAT_ABI_MASK) == EF_LOONGARCH_FLOAT_ABI_SINGLE)
+  (!((EF_LOONGARCH_ABI(abi) & EF_LOONGARCH_ABI_FLOAT_MASK) ^ EF_LOONGARCH_ABI_SINGLE_FLOAT_MASK))
 #define EF_LOONGARCH_IS_DOUBLE_FLOAT(abi) \
-  ((abi & EF_LOONGARCH_FLOAT_ABI_MASK) == EF_LOONGARCH_FLOAT_ABI_DOUBLE)
+  (!((EF_LOONGARCH_ABI(abi) & EF_LOONGARCH_ABI_FLOAT_MASK) ^ EF_LOONGARCH_ABI_DOUBLE_FLOAT_MASK))
 
-#define EF_LOONGARCH_ABI (EF_LOONGARCH_ABI_MASK | EF_LOONGARCH_FLOAT_ABI_MASK)
 
 #endif /* _ELF_LOONGARCH_H */
-- 
2.34.0

